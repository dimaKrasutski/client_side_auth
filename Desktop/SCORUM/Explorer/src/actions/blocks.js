import { createActions } from 'redux-actions';

export const {
  setBlocks,
  setBlockNumber,
  setCurrentBlockNumber,
} = createActions(
  'SET_BLOCKS',
  'SET_BLOCK_NUMBER',
  'SET_CURRENT_BLOCK_NUMBER',
);
