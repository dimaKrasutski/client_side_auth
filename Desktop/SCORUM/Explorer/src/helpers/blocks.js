import { compose, last, dec, head, reverse } from 'ramda';
import { Subject } from 'rxjs/Subject';
import { getBlocksHistory } from '../api/scorum';

/**
 * Get first transaction from a list and take index property decrementing it
 */
export const getNextOffset = compose(dec, head, last);

/**
 * @param {function} reducer - function to reduce entity items
 * @param {number} blockNumber - block number
 * @param {function} setCurrentBlock - set current block
 */
export const createBlocksFetcher = (reducer, blockNumber, setCurrentBlock) => {
  const transactionsSubject = new Subject();
  let blockNum = blockNumber;
  let isStopped = false;
  let inProgress = false;

  transactionsSubject.hasNext = () => !isStopped;

  transactionsSubject.fetchNext = function fetchNext() {
    if (isStopped || inProgress) {
      return null;
    }

    const getBlocks = getBlocksHistory(blockNum);
    inProgress = true;
    getBlocks.subscribe(compose(
      (fetched) => {
        this.next(fetched);
        blockNum = getNextOffset(fetched);
        setCurrentBlock(blockNum);
        if (Number.isNaN(blockNum) || blockNum < 0) { // When we trap to negative - no more records left
          isStopped = true;
        }
        inProgress = false;
      },
      reverse,
    ));

    return null;
  };

  return transactionsSubject;
};
