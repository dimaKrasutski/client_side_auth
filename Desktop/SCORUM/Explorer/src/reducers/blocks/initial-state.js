const initialState = {
  blocks: [],
  currentBlockNumber: -1,
  headBlockNumber: -1,
};

export default initialState;
