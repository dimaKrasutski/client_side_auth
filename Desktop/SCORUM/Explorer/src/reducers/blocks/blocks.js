import { handleActions } from 'redux-actions';
import initialState from './initial-state';
import { setBlocks, setBlockNumber, setCurrentBlockNumber } from '../../actions/blocks';

export default handleActions(
  {
    [setBlocks]: (state, { payload: blocks }) => ({
      ...state,
      blocks: state.blocks.concat(blocks),
    }),
    [setBlockNumber]: (state, { payload: props }) => ({
      ...state,
      headBlockNumber: props.head_block_number,
      currentBlockNumber: props.head_block_number,
    }),
    [setCurrentBlockNumber]: (state, { payload: number }) => ({
      ...state,
      currentBlockNumber: number,
    }),
  },
  initialState,
);
