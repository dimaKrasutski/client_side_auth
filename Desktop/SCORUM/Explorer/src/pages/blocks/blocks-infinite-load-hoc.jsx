import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'ramda';
import { createBlocksFetcher } from '../../helpers/blocks';
import { withInfiniteScroll } from '../../hocs/with-infinite-scroll';
import { setBlocks, setCurrentBlockNumber } from '../../actions/blocks';

export const withBlocksInfiniteLoad = (WrappedComponent, reducer) => {
  const WrappedWithInfiniteScroll = withInfiniteScroll(WrappedComponent);

  class BlocksInfiniteLoad extends Component {
    static propTypes = {
      currentBlockNumber: PropTypes.number.isRequired,
      headBlockNumber: PropTypes.number.isRequired,
      setBlocks: PropTypes.func.isRequired,
      setCurrentBlockNumberAction: PropTypes.func.isRequired,
    };

    blocksFetcher = createBlocksFetcher(
      reducer,
      this.props.currentBlockNumber,
      this.props.setCurrentBlockNumberAction
    );

    blocksFetcherSubscription;

    state = {
      inProgress: false,
    };

    /**
     * Initial subscribe to fetcher and instant preload of first bunch
     */
    componentDidMount() {
      const { currentBlockNumber, headBlockNumber } = this.props;
      this.blocksFetcherSubscription = this.blocksFetcher.subscribe(this.updateBlocks);
      if (currentBlockNumber === headBlockNumber) {
        this.blocksFetcher.fetchNext();
      }
    }

    componentWillUnmount() {
      this.blocksFetcherSubscription.unsubscribe();
    }

    /**
     * Initiate loading of next transactions
     */
    loadNext = () => {
      if (this.blocksFetcher.hasNext()) {
        this.setState({
          inProgress: true,
        });
        this.blocksFetcher.fetchNext();
      }
    };

    updateBlocks = (nextBlocks) => {
      this.props.setBlocks(nextBlocks);
      this.setState({
        inProgress: false,
      });
    };

    render() {
      const { inProgress, blocks } = this.state;

      return (
        <WrappedWithInfiniteScroll
          {...this.props}
          inProgress={inProgress}
          loadNext={this.loadNext}
          blocks={blocks}
        />
      );
    }
  }

  const mapStateToProps = ({ blocks }) => ({
    currentBlockNumber: blocks.currentBlockNumber,
    headBlockNumber: blocks.headBlockNumber,
  });

  const mapDispatchToProps = dispatch => ({
    setBlocks: compose(dispatch, setBlocks),
    setCurrentBlockNumberAction: compose(dispatch, setCurrentBlockNumber),
  });

  return connect(mapStateToProps, mapDispatchToProps)(BlocksInfiniteLoad);
};
