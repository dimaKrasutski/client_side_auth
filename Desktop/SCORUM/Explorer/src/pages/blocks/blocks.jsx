import React from 'react';
import PropTypes from 'prop-types';
import { Container, Row, Col } from 'reactstrap';
import { connect } from 'react-redux';
import { Main } from '../main';
import { BlockTable } from './blocks-table/blocks-table';
import { PageTitle } from '../../components/page-title/page-title';
import { withBlocksInfiniteLoad } from './blocks-infinite-load-hoc';

const BlocksWithInfiniteLoad = withBlocksInfiniteLoad(BlockTable, null);

export const BlocksUI = ({ currentBlockNumber }) => (
  <Main>
    <Container className="pt-6 blocks-container">
      <Row>
        <Col xs="12">
          <PageTitle title="explorer.blocks.title" />
          {
            !!currentBlockNumber &&
            <BlocksWithInfiniteLoad />
          }
        </Col>
      </Row>
    </Container>
  </Main>
);

BlocksUI.propTypes = {
  currentBlockNumber: PropTypes.number.isRequired,
};

const mapStateToProps = ({ blocks }) => ({
  currentBlockNumber: blocks.currentBlockNumber,
});

export const Blocks = connect(mapStateToProps)(BlocksUI);
