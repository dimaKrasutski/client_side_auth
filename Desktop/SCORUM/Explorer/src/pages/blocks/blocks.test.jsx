import React from 'react';
import { BlocksUI } from './blocks';

describe('<BlocksUI />', () => {
  test('should match snapshot', () => {
    const wrapper = shallow(<BlocksUI currentBlockNumber="-1" />);
    expect(wrapper).toMatchSnapshot();
  });
});
