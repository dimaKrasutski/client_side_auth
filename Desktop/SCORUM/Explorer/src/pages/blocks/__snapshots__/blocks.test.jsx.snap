// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlocksUI /> should match snapshot 1`] = `
<Connect(MainUI)>
  <Container
    className="pt-6 blocks-container"
    tag="div"
  >
    <Row
      tag="div"
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xs="12"
      >
        <PageTitle
          secondaryTitle=""
          title="explorer.blocks.title"
        />
        <Connect(BlocksInfiniteLoad) />
      </Col>
    </Row>
  </Container>
</Connect(MainUI)>
`;
